tools/ci-build/runtime-versioner/src/index.rs

```diff
@@ -0,0 +1,102 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+use anyhow::{anyhow, Context, Error, Result};
+use camino::Utf8Path;
+use crates_index::Crate;
+use reqwest::StatusCode;
+use std::collections::HashMap;
+use std::fs;
+
+pub struct CratesIndex(Inner);
+
+enum Inner {
+    Fake(FakeIndex),
+    Real(crates_index::SparseIndex),
+}
+
+impl CratesIndex {
+    /// Returns a real sparse crates.io index.
+    pub fn real() -> Result<Self> {
+        Ok(Self(Inner::Real(
+            crates_index::SparseIndex::new_cargo_default()
+                .context("failed to initialize the sparse crates.io index")?,
+        )))
+    }
+
+    /// Returns a fake crates.io index from file, panicking if loading fails.
+    pub fn fake(path: impl AsRef<Utf8Path>) -> Self {
+        Self(Inner::Fake(FakeIndex::from_file(path)))
+    }
+
+    /// Retrieves the published versions for the given crate name.
+    pub fn published_versions(&self, crate_name: &str) -> Result<Vec<String>> {
+        match &self.0 {
+            Inner::Fake(index) => Ok(index.crates.get(crate_name).cloned().unwrap_or_default()),
+            Inner::Real(index) => {
+                let url = index
+                    .crate_url(crate_name)
+                    .expect("crate name is not empty string");
+                let crate_meta: Option<Crate> = reqwest::blocking::get(url)
+                    .map_err(Error::from)
+                    .and_then(|response| {
+                        let status = response.status();
+                        response.bytes().map(|b| (status, b)).map_err(Error::from)
+                    })
+                    .and_then(|(status, bytes)| match status {
+                        status if status.is_success() => {
+                            Crate::from_slice(&bytes).map_err(Error::from).map(Some)
+                        }
+                        StatusCode::NOT_FOUND => Ok(None),
+                        status => {
+                            let body = String::from_utf8_lossy(&bytes);
+                            Err(anyhow!(
+                                "request to crates.io index failed ({status}):\n{body}"
+                            ))
+                        }
+                    })
```

Comment:
is this going to be run by humans or in CI? may want retries at some layer—crates.io can often be down. (Or is this hitting GitHub, same thing...). 

if it's hitting GitHub, unauthenticated requests are often throttled very quickly

Reply:
Good call. Will need to do some refactoring to pull out the retry logic in the publisher tool into common.